src/java.base/share/classes/jdk/crac/Core.java

```diff
@@ -178,11 +181,20 @@ private static void checkpointRestore1() throws
     public static void checkpointRestore() throws
             CheckpointException,
             RestoreException {
-        try {
-            checkpointRestore1();
-        } finally {
-            if (traceStartupTime) {
-                System.out.println("STARTUPTIME " + System.nanoTime() + " restore-finish");
+        synchronized (checkpointRestoreLock) {
+            if (!checkpointInProgress) {
+                try {
+                    checkpointInProgress = true;
```

Comment:
Probably worth a comment that this guards from recursive checkpoints called from Resource's methods that will successfully enter the synchronized section -- just because the handling of the checkpointInProgress is below in the text

Reply:
Added comments